### PR TITLE
Added Tribler twistd plugin

### PR DESCRIPTION
**Pull request by devos50**
This PR adds a twistd plugin so Tribler can be started using the `twistd` command. To run the plugin, `cd` to the root folder of Tribler (not the `Tribler` directory) and execute: `PYTHONPATH=twisted twistd -n tribler`. This will run the twistd daemon.

Note that this PR requires https://github.com/Tribler/dispersy/pull/474 to work (otherwise, Tribler hangs during startup).

@whirm Should I cover the code in this PR with a test? I'm not sure how I should write a test for this plugin.

### CODE REVIEW

**Comment by devos50** [comment]
@lfdversluis can you review?

File: twisted/twisted/plugins/tribler_plugin.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        """
+
+        def signal_handler(sig, _):
+            msg("Received shut down signal %s" % sig)
+            if not self._stopping:
+                self._stopping = True
+                try:
+                    self.session.shutdown()
+                    msg("Tribler shut down")
+                except Exception, exc:
```
Comment (lfdversluis) [inline twisted/twisted/plugins/tribler_plugin.py]: general exception, can you replace this with a tuple?

File: twisted/twisted/plugins/tribler_plugin.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            port = options["manhole"]
+            manhole = manhole_tap.makeService({
+                'namespace': manhole_namespace,
+                'telnetPort': 'tcp:%d:interface=127.0.0.1' % port,
+                'sshPort': None,
+                'passwd': os.path.join(os.path.dirname(__file__), 'passwd'),
+            })
+            tribler_service.addService(manhole)
+
+        reactor.exitCode = 0
```
Comment (lfdversluis) [inline twisted/twisted/plugins/tribler_plugin.py]: Why this statement?

File: twisted/twisted/plugins/tribler_plugin.py (hunk 1 above)
Comment (devos50) [inline twisted/twisted/plugins/tribler_plugin.py]: I've removed the try/catch block. We are not catching exceptions during Tribler shutdown in other files (`tribler_main.py`, `test_as_server.py`, `tunnel/main.py` etc).

File: twisted/twisted/plugins/tribler_plugin.py (hunk 2 above)
Comment (devos50) [inline twisted/twisted/plugins/tribler_plugin.py]: I copied most of the file from the Dispersy twistd plugin and this line got copied over. I removed it, tested the shutdown procedure and it seems that the statement is indeed not necessary.

**Comment by devos50** [comment]
@whirm can you review?

**Comment by whirm** [comment]
re. tests: Have a look at how the twisted guys test their stuff, maybe you can get some ideas.

File: twisted/twisted/plugins/tribler_plugin.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class Options(usage.Options):
+    optParameters = [
+        ["manhole", "m", 0, "Enable manhole telnet service listening at the specified port", int],
+    ]
+
+
+class TriblerServiceMaker(object):
+    implements(IServiceMaker, IPlugin)
+    tapname = "tribler"
+    description = "Tribler twistd plugin, starts Tribler as a twistd process"
```
Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: "as a service"

File: twisted/twisted/plugins/tribler_plugin.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    options = Options
+
+    def __init__(self):
+        """
+        Initialize the variables of the TriblerServiceMaker and the logger.
+        """
+        self.session = None
+        self._stopping = False
+        self._logger = logging.getLogger(self.__class__.__name__)
```
Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: This is not being used

File: twisted/twisted/plugins/tribler_plugin.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        signal.signal(signal.SIGINT, signal_handler)
+        signal.signal(signal.SIGTERM, signal_handler)
+
+        msg("Starting Tribler")
+
+        config = SessionStartupConfig()
+        config.set_http_api_enabled(True)
+        self.session = Session(config)
+        upgrader = self.session.prestart()
+        while not upgrader.is_done:
```
Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: upgrader.run() is blocking, so this while loop does nothing. Also, you are not checking if the upgrade failed.

Another issue we probably have is that the notifications are going to pile up until the upgrade process is finished (AFAIR notify schedules the notifiation stuff to be called on the thread pool which will be scheduling stuff to be called on the reactor thread, which (as the upgrade is blocking ATM) will be busy until the upgrade finishes and then all the notifications the upgrader generated will be delivered (to no one until you implement it on the REST interface). So remember that we need to "twistify" the upgrade code before implementing the REST notifications for it.

Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: re. not checking if the upgrade failed: I know the upgrader will move the .Tribler dir away and start from scratch, but do we really want this to happen on a headless server?

Comment (devos50) [inline twisted/twisted/plugins/tribler_plugin.py]: If the upgrader fails, I think we need to do something to make sure Tribler starts up. Starting with a fresh .Tribler directory is the best choice probably, whether we are running headless or not.

Are there any other advantages of not starting from scratch when running headless?

Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: The difference is that when it's running headless, the user won't know if he doesn't check the log file immediately after starting the service. I guess we should make it fail so the user notices.

**Comment by devos50** [comment]
@whirm see my last fixup commit where I made the plugin fail if upgrading fails. I also updated the dispersy pointer.

Also, I'm unable to find tests in the Twisted source code for their shipped plugins, only for their core functionality. Also, it lives outside Tribler and it's only a utility script to startup Tribler so I don't think it really needs a dedicated test case.

File: twisted/twisted/plugins/tribler_plugin.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            port = options["manhole"]
+            manhole = manhole_tap.makeService({
+                'namespace': manhole_namespace,
+                'telnetPort': 'tcp:%d:interface=127.0.0.1' % port,
+                'sshPort': None,
+                'passwd': os.path.join(os.path.dirname(__file__), 'passwd'),
+            })
+            tribler_service.addService(manhole)
+
+        reactor.callWhenRunning(self.start_tribler)
```
Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: I'm thinking this is not needed for a normal reactor, you should just do everything here as the reactor will be running.

File: twisted/twisted/plugins/tribler_plugin.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        msg("Starting Tribler")
+
+        config = SessionStartupConfig()
+        config.set_http_api_enabled(True)
+        self.session = Session(config)
+        upgrader = self.session.prestart()
+        if upgrader.failed:
+            msg("The upgrader failed: .Tribler directory backed up, aborting")
+            reactor.stop()
```
Comment (whirm) [inline twisted/twisted/plugins/tribler_plugin.py]: You shouldn't stop the reactor, that's twistd's job.

If raising an exception once you have inlined this with `makeService()` doesn't make twistd exit with nonzero status, then you could do something like this:

```python
reactor.addSystemEventTrigger('after', 'shutdown', os._exit, EXIT_CODE)
```

File: twisted/twisted/plugins/tribler_plugin.py (hunk 6 above)
Comment (devos50) [inline twisted/twisted/plugins/tribler_plugin.py]: The reactor is not running when I schedule the `self.start_tribler` call (tested with `reactor.running`).

**Comment by devos50** [comment]
@whirm see my last fixup commit. I've also updated the dispersy pointer to include the netifaces fix.

**Comment by whirm** [comment]
Ok, squash it

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
(anon tunnel tests timeout)

**Comment by devos50** [comment]
@whirm ready

**Pull request merged**